### PR TITLE
Add csharp_configure rule

### LINKED ISSUE #1: mcs-finding should be more robust
See http://ci.bazel.io/job/rules_dotnet/BAZEL_VERSION=HEAD,PLATFORM_NAME=linux-x86_64/3/console:

```
____[7 / 11] Compiling examples/example_transitive_lib:TransitiveClass
ERROR: /home/ci/workspace/rules_dotnet/BAZEL_VERSION/HEAD/PLATFORM_NAME/linux-x86_64/examples/example_transitive_lib/BUILD:3:1: error executing shell command: '/usr/local/bin/mcs -out:bazel-out/local_linux-fastbuild/bin/examples/example_transitive_lib/TransitiveClass.dll -target:library -fullpaths -warn:4 -nologo -doc:bazel-out/local_linux-fastbuild/bin/e...' failed: namespace-sandbox failed: error executing command 
  (cd /home/ci/.cache/bazel/_bazel_ci/1471c2a9e7dce6d4e1d950b43de46d01/linux-x86_64 && \
  exec env - \
  /home/ci/.cache/bazel/_bazel_ci/1471c2a9e7dce6d4e1d950b43de46d01/linux-x86_64/_bin/namespace-sandbox @/home/ci/.cache/bazel/_bazel_ci/1471c2a9e7dce6d4e1d950b43de46d01/linux-x86_64/bazel-sandbox/e5591ace-6edb-42f2-b6ab-9581eb783b26-0.params -- /bin/bash -c '/usr/local/bin/mcs -out:bazel-out/local_linux-fastbuild/bin/examples/example_transitive_lib/TransitiveClass.dll -target:library -fullpaths -warn:4 -nologo -doc:bazel-out/local_linux-fastbuild/bin/examples/example_transitive_lib/TransitiveClass.xml -debug examples/example_transitive_lib/Properties/AssemblyInfo.cs examples/example_transitive_lib/TransitiveClass.cs' '' -out:bazel-out/local_linux-fastbuild/bin/examples/example_transitive_lib/TransitiveClass.dll -target:library -fullpaths -warn:4 -nologo -doc:bazel-out/local_linux-fastbuild/bin/examples/example_transitive_lib/TransitiveClass.xml -debug).
: /usr/local/bin/mcs: No such file or directory
```

Right now /usr/local/bin/mcs is hardcoded and I'm not sure it's included in the sandbox.

### PR DESCRIPTION
**Pull request by davidzchen**
Add csharp_configure rule to find and use the Mono C# toolchain installed on the local system.

Also, add inline documentation for rules and macros.

Fixes #1

### CODE REVIEW

**Comment by davidzchen** [comment]
+cc @damienmg, @kchodorow

**Comment by damienmg** [comment]
LGTM

I don't think this is going to fix the build anyway because I don't think we have mono installed.

**Comment by damienmg** [comment]
ping @googlebot

**Comment by damienmg** [comment]
(@davidzchen: can you try to ping @googlebot to test?)

**Comment by davidzchen** [comment]
ping @googlebot

File: dotnet/csharp.bzl [hunk 1]
```diff
@@ -232,11 +242,38 @@ def _cs_nunit_run_impl(ctx):
...
                 runfiles=runfiles)
 
+def _find_and_symlink(repository_ctx, binary, env_variable):
+  if env_variable in repository_ctx.os.environ:
+    return repository_ctx.path(repository_ctx.os.environ[env_variable])
+  else:
+    found_binary = repository_ctx.which(binary)
+    if found_binary == None:
+      fail("Cannot find %s. Either correct your path or set the %s " +
+           "environment variable.")
```
Comment (damienmg) [inline dotnet/csharp.bzl]: missing % (binary, env_variable)

**Pull request merged**